vendor/bat-native-ads/src/bat/ads/internal/classification/purchase_intent_classifier/purchase_intent_classifier.cc

```diff
@@ -112,5 +122,267 @@ uint16_t PurchaseIntentClassifier::GetIntentScoreForHistory(
   return intent_score;
 }
 
+SiteInfo PurchaseIntentClassifier::GetSite(
+    const std::string& url) {
+  const GURL visited_url = GURL(url);
+  SiteInfo info;
+
+  if (!visited_url.has_host()) {
+    return info;
+  }
+
+  for (const auto& site : sites_) {
+    const GURL site_url = GURL(site.url_netloc);
+
+    if (!site_url.is_valid()) {
+      continue;
+    }
+
+    if (SameDomainOrHost(visited_url, site_url,
+        net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES)) {
+      info = site;
+      return info;
+    }
+  }
+
+  return info;
+}
+
+PurchaseIntentSegmentList PurchaseIntentClassifier::GetSegments(
+    const std::string& search_query) {
+  PurchaseIntentSegmentList segment_list;
+  auto search_query_keyword_set = TransformIntoSetOfWords(search_query);
+
+  for (const auto& keyword : segment_keywords_) {
+    auto list_keyword_set = TransformIntoSetOfWords(keyword.keywords);
+
+    // Intended behaviour relies on early return from list traversal and
+    // implicitely on the ordering of |segment_keywords_| to ensure
+    // specific segments are matched over general segments, e.g. "audi a6"
+    // segments should be returned over "audi" segments if possible.
+    if (IsSubset(search_query_keyword_set, list_keyword_set)) {
+      segment_list = keyword.segments;
+      return segment_list;
+    }
+  }
+
+  return segment_list;
+}
+
+uint16_t PurchaseIntentClassifier::GetFunnelWeight(
+    const std::string& search_query) {
+  auto search_query_keyword_set = TransformIntoSetOfWords(search_query);
+
+  uint16_t max_weight = kPurchaseIntentDefaultSignalWeight;
+  for (const auto& keyword : funnel_keywords_) {
+    auto list_keyword_set = TransformIntoSetOfWords(keyword.keywords);
+
+    if (IsSubset(search_query_keyword_set, list_keyword_set) &&
+        keyword.weight > max_weight) {
+      max_weight = keyword.weight;
+    }
+  }
+
+  return max_weight;
+}
+
+// TODO(Moritz Haller): Rename to make explicit that method checks set_a is
+// subset of set_b but not vice versa
+// TODO(https://github.com/brave/brave-browser/issues/8495): Implement Brave
+// Ads Purchase Intent keyword matching with std::sets
+bool PurchaseIntentClassifier::IsSubset(
+    std::vector<std::string> keyword_set_a,
```

Comment:
Would be better to pass these by reference or const reference if possible./

Reply:
Fixed